File: src/main/java/com/github/unchama/contextualexecutor/builder/ArgumentParserScope.kt
package com.github.unchama.contextualexecutor.builder

import arrow.core.Left
import arrow.core.Right
import com.github.unchama.effect.*
import org.bukkit.command.CommandSender

/**
 * [ContextualExecutorBuilder.argumentsParser]が要求する,
 * 引数文字列から[ResponseEffectOrResult<Any>]への関数の作成を行うためのスコープオブジェクト.
 *
 * [ArgumentParserScope.ScopeProvider.parser]を通してスコープ付き関数をそのような関数に変換できる.
 */
object ArgumentParserScope {
  fun <CS> failWith(message: TargetedEffect<CS>): ResponseEffectOrResult<CS, Nothing> = Left(message)

  /**
   * メッセージなしで「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  fun failWithoutError(): ResponseEffectOrResult<Any?, Nothing> = failWith(EmptyEffect)

  /**
   * メッセージ付きの「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  fun failWith(message: String): ResponseEffectOrResult<CommandSender, Nothing> = failWith(message.asMessageEffect())

  /**
   * メッセージ付きの「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  fun failWith(message: List<String>): ResponseEffectOrResult<CommandSender, Any> = failWith(message.asMessageEffect())

  /**
   * [result]により「成功」したことを示す[ResponseEffectOrResult]を作成する.
   */
  fun succeedWith(result: Any): ResponseEffectOrResult<Any?, Any> = Right(result)

  object ScopeProvider {
    /**
     * [ArgumentParserScope]のスコープ付き関数をプレーンな関数へと変換する.
     */
    fun <CS> parser(
        parse: ArgumentParserScope.(String) -> ResponseEffectOrResult<CS, Any>
    ): (String) -> ResponseEffectOrResult<CS, Any> = { argument -> with(ArgumentParserScope) { parse(argument) } }
  }
}

File: src/main/scala/com/github/unchama/contextualexecutor/builder/ArgumentParserScope.scala
package com.github.unchama.contextualexecutor.builder

import com.github.unchama.targetedeffect.TargetedEffect.emptyEffect
import com.github.unchama.targetedeffect._
import com.github.unchama.targetedeffect.player.MessageEffect
import org.bukkit.command.CommandSender

/**
 * 文字列から`ResponseEffectOrResult[Any]`への関数の作成を行うためのオブジェクト.
 *
 * [ArgumentParserScope.ScopeProvider.parser]を通してスコープ付き関数をそのような関数に変換できる.
 */
// TODO update comment
object ArgumentParserScope {
  /**
   * メッセージなしで「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  def failWithoutError(): ResponseEffectOrResult[Any, Nothing] = failWith(emptyEffect)

  /**
   * メッセージ付きの「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  def failWith(message: String): ResponseEffectOrResult[CommandSender, Nothing] = failWith(MessageEffect(message))

  def failWith[CS](effect: TargetedEffect[CS]): ResponseEffectOrResult[CS, Nothing] = Left(effect)

  /**
   * メッセージ付きの「失敗」を表す[ResponseEffectOrResult]を作成する.
   */
  def failWith(message: List[String]): ResponseEffectOrResult[CommandSender, Any] = failWith(MessageEffect(message))

  /**
   * [result]により「成功」したことを示す[ResponseEffectOrResult]を作成する.
   */
  def succeedWith(result: Any): ResponseEffectOrResult[Any, Any] = Right(result)
}
